From fe0b517fa34063990a83268edf7a3cd9ba0b2362 Mon Sep 17 00:00:00 2001
From: Yuchen Wu <yuchen@cloudflare.com>
Date: Mon, 13 Mar 2023 14:28:10 -0700
Subject: [PATCH] PINGORA-474: disable cert version check for openssl
 compatibility

Some free customers use badly crafted cert (with verification off).
Openssl allows these cert but boring does't.
---
 crypto/x509/x509_test.cc | 6 +++---
 crypto/x509/x_crl.cc     | 8 +++-----
 crypto/x509/x_x509.cc    | 7 ++++---
 3 files changed, 10 insertions(+), 11 deletions(-)

diff --git a/crypto/x509/x509_test.cc b/crypto/x509/x509_test.cc
index e5d64a9c3..34bd80e64 100644
--- a/crypto/x509/x509_test.cc
+++ b/crypto/x509/x509_test.cc
@@ -4430,11 +4430,11 @@ TEST(X509Test, InvalidVersion) {
   EXPECT_FALSE(CertFromPEM(kNegativeVersionPEM));
   EXPECT_FALSE(CertFromPEM(kFutureVersionPEM));
   EXPECT_FALSE(CertFromPEM(kOverflowVersionPEM));
-  EXPECT_FALSE(CertFromPEM(kV1WithExtensionsPEM));
-  EXPECT_FALSE(CertFromPEM(kV2WithExtensionsPEM));
+  EXPECT_TRUE(CertFromPEM(kV1WithExtensionsPEM));
+  EXPECT_TRUE(CertFromPEM(kV2WithExtensionsPEM));
   EXPECT_FALSE(CertFromPEM(kV1WithIssuerUniqueIDPEM));
   EXPECT_FALSE(CertFromPEM(kV1WithSubjectUniqueIDPEM));
-  EXPECT_FALSE(CRLFromPEM(kV1CRLWithExtensionsPEM));
+  EXPECT_TRUE(CRLFromPEM(kV1CRLWithExtensionsPEM));
   EXPECT_FALSE(CRLFromPEM(kV1CRLWithEntryExtensionsPEM));
   EXPECT_FALSE(CRLFromPEM(kV3CRLPEM));
   EXPECT_FALSE(CSRFromPEM(kV2CSRPEM));
diff --git a/crypto/x509/x_crl.cc b/crypto/x509/x_crl.cc
index d30c17253..0830ad348 100644
--- a/crypto/x509/x_crl.cc
+++ b/crypto/x509/x_crl.cc
@@ -156,11 +156,9 @@ static int crl_cb(int operation, ASN1_VALUE **pval, const ASN1_ITEM *it,
         }
       }
 
-      // Per RFC 5280, section 5.1.2.1, extensions require v2.
-      if (version != X509_CRL_VERSION_2 && crl->crl->extensions != nullptr) {
-        OPENSSL_PUT_ERROR(X509, X509_R_INVALID_FIELD_FOR_VERSION);
-        return 0;
-      }
+      // OpenSSL compatibility: accept extensions in v1 CRLs, although RFC
+      // 5280, section 5.1.2.1 requires v2.
+      // https://datatracker.ietf.org/doc/html/rfc5280#section-5.1.2.1
 
       // Extensions is a SEQUENCE SIZE (1..MAX), so it cannot be empty. An empty
       // extensions list is encoded by omitting the OPTIONAL field.
diff --git a/crypto/x509/x_x509.cc b/crypto/x509/x_x509.cc
index 21f53803a..a6ebf4dde 100644
--- a/crypto/x509/x_x509.cc
+++ b/crypto/x509/x_x509.cc
@@ -151,9 +151,10 @@ X509 *X509_parse_with_algorithms(CRYPTO_BUFFER *buf,
       return nullptr;
     }
   }
-  // Per RFC 5280, section 4.1.2.9, extensions require v3:
-  if (ret->version >= X509_VERSION_3 &&
-      CBS_peek_asn1_tag(&tbs, kExtensionsTag)) {
+  // OpenSSL compatibility: accept extensions in v1 and v2 certificates,
+  // although RFC 5280, section 4.1.2.9 requires v3.
+  // https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.9
+  if (CBS_peek_asn1_tag(&tbs, kExtensionsTag)) {
     CBS wrapper;
     if (!CBS_get_asn1(&tbs, &wrapper, kExtensionsTag)) {
       OPENSSL_PUT_ERROR(ASN1, ASN1_R_DECODE_ERROR);
-- 
2.39.5
